File: NOIP/Final/NOIP2010/SeniorRefCode/Pascal/flow.pas
program flow(input,output);
  const
    maxn=500;
    dx:array[1..4] of longint=(0,0,1,-1);
    dy:array[1..4] of longint=(1,-1,0,0);
  var
    n,m,ans,i,j:longint;
    q:array[1..maxn*maxn*2,1..2] of longint;
    s,h:array[1..maxn,1..maxn] of longint;
    st,ed,dp:array[0..maxn] of longint;
    flag:boolean;

  procedure judge;
    var
      f,r,x,y,tx,ty,k,i:longint;
    begin
      fillchar(s,sizeof(s),0);
      for k:=1 to m do if s[1,k]=0 then begin
        f:=1;r:=0;
        inc(r);q[r,1]:=1;q[r,2]:=k;s[1,k]:=1;
        while f<=r do begin
          x:=q[f,1];y:=q[f,2];inc(f);
          for i:=1 to 4 do begin
            tx:=x+dx[i];ty:=y+dy[i];
            if (tx>=1)and(tx<=n)and(ty>=1)and(ty<=m) then
              if (s[tx,ty]=0)and(h[tx,ty]<h[x,y]) then begin
                inc(r);
                q[r,1]:=tx;q[r,2]:=ty;
                s[tx,ty]:=1;
              end;
          end;
        end;
      end;
      flag:=true;ans:=0;
      for i:=1 to m do
        if s[n,i]=0 then begin
          flag:=false;inc(ans);
        end;
    end;

  procedure work;
    var
      f,r,x,y,tx,ty,k,i,j,min:longint;
    begin
      for k:=1 to m do begin
        f:=1;r:=0;fillchar(s,sizeof(s),0);
        inc(r);q[r,1]:=1;q[r,2]:=k;s[1,k]:=1;
        while f<=r do begin
          x:=q[f,1];y:=q[f,2];inc(f);
          for i:=1 to 4 do begin
            tx:=x+dx[i];ty:=y+dy[i];
            if (tx>=1)and(tx<=n)and(ty>=1)and(ty<=m) then
              if (s[tx,ty]=0)and(h[tx,ty]<h[x,y]) then begin
                inc(r);
                q[r,1]:=tx;q[r,2]:=ty;
                s[tx,ty]:=1;
              end;
          end;
        end;
        for i:=1 to m do
          if s[n,i]<>0 then begin
            st[k]:=i;break;
          end;
        for i:=m downto 1 do
          if s[n,i]<>0 then begin
            ed[k]:=i;break;
          end;
      end;
      fillchar(dp,sizeof(dp),0);
      dp[0]:=0;
      for i:=1 to m do begin
        min:=maxlongint;
        for j:=1 to m do if (st[j]<=i)and(ed[j]>=i) then
          if dp[st[j]-1]<min then min:=dp[st[j]-1];
        dp[i]:=min+1;
      end;
      ans:=dp[m];
    end;

  begin
    assign(input,'flow.in');reset(input);
    assign(output,'flow.out');rewrite(output);
    readln(n,m);
    for i:=1 to n do begin
      for j:=1 to m do
        read(h[i,j]);
      readln;
    end;
    judge;
    if flag then begin
      work;
      writeln(1);
      writeln(ans);
    end
    else begin
      writeln(0);
      writeln(ans);
    end;
    close(input);close(output);
  end.

File: NOIP/Repecharge/NOIP2010/SeniorRefCode/C++/flow.cpp
#include <iostream>
#include <fstream>
using namespace std;

const int maxn=500;const int dx[4]={1,-1,0,0};
const int dy[4]={0,0,1,-1};
int n,m,ans;
int q[maxn*maxn*2][2],s[maxn+1][maxn+1],h[maxn+1][maxn+1];
int st[maxn+1],ed[maxn+1],dp[maxn+1];
bool flag;
void judge()
{
	int f,r;
	int x,y,tx,ty;
	memset(s,0,sizeof(s));
	for(int k=1;k<=m;k++) if(s[1][k]==0){
		f=1;r=0;
		r++;q[r][0]=1;q[r][1]=k;s[1][k]=1;
		while(f<=r){
			x=q[f][0];y=q[f][1];f++;
			for(int i=0;i<4;i++){
				tx=x+dx[i];ty=y+dy[i];
				if((tx>=1)&&(tx<=n)&&(ty>=1)&&(ty<=m)) 
					if((s[tx][ty]==0)&&(h[tx][ty]<h[x][y])){
						r++;q[r][0]=tx;q[r][1]=ty;s[tx][ty]=1;
					}
			}
		}
	}
	flag=true;ans=0;
	for(int i=1;i<=m;i++)
		if(s[n][i]==0){
			flag=false;ans++;
		}
}

void work(){
	int x,y,tx,ty,f,r;
	for(int k=1;k<=m;k++){
		f=1;r=0;
		memset(s,0,sizeof(s));
		r++;q[r][0]=1;q[r][1]=k;s[1][k]=1;
		while(f<=r){
			x=q[f][0];y=q[f][1];f++;
			for(int i=0;i<4;i++){
				tx=x+dx[i];ty=y+dy[i];
				if((tx>=1)&&(tx<=n)&&(ty>=1)&&(ty<=m)) if((s[tx][ty]==0)&&(h[tx][ty]<h[x][y])){
					r++;q[r][0]=tx;q[r][1]=ty;s[tx][ty]=1;
				}
			}
		}
		for(int i=1;i<=m;i++)
			if(s[n][i]!=0){st[k]=i;break;}
		for(int i=m;i>=1;i--)
			if(s[n][i]!=0){ed[k]=i;break;}
	}
	memset(dp,0,sizeof(dp));
	dp[0]=0;
	int min;
	for(int i=1;i<=m;i++){
		min=100000000;
		for(int j=1;j<=m;j++) if((st[j]<=i)&&(ed[j]>=i))
			if(dp[st[j]-1]<min) min=dp[st[j]-1];
		dp[i]=min+1;
	}
	ans=dp[m];
}

int main()
{
	ifstream infile("flow.in",ios::in);
	ofstream outfile("flow.out",ios::out);
	infile>>n>>m;
	for(int i=1;i<=n;i++)
		for(int j=1;j<=m;j++)
			infile>>h[i][j];
	judge();
	if(flag==false){
		outfile<<0<<endl;
		outfile<<ans<<endl;
	}
	else{
		work();
		outfile<<1<<endl;
		outfile<<ans<<endl;
	}
	infile.close();
	outfile.close();
	return 0;
}
